db/migrate/20190617191109_move_awx_credentials_to_authentications.rb

```diff
@@ -0,0 +1,327 @@
+    serialize :options
+  end
+
+  # This class is port of the code found in both the awx code base, and the
+  # python cryptography library:
+  #
+  #   https://github.com/pyca/cryptography/blob/ed5a7717/src/cryptography/fernet.py
+  #   https://github.com/ansible/awx/blob/4788f0814/awx/main/utils/encryption.py
+  #
+  # The only real portions from this that are from the `awx` side are for the
+  # inheriting `Fernet256` from the pythons cryptography lib's `Fernet` class:
+  #
+  #   https://github.com/ansible/awx/blob/4788f081/awx/main/utils/encryption.py#L20-L23
+  #
+  # The rest pretty much emulates code from cryptography.py
+  class Fernet256
+    attr_reader :hmac, :cipher, :signing_key, :encryption_key
+
+    InvalidToken = Class.new(ArgumentError)
+
+    # This initialize method is mostly pulled from from the ansible override of
+    # the base Fernet library.
+    #
+    #   https://github.com/ansible/awx/blob/4788f0814/awx/main/utils/encryption.py#L24-L35
+    #
+    # The backend is always going to be `OpenSSL` for us, so that is
+    # irrelevant, but everything else is mostly as it is on the python
+    # equivalent.
+    #
+    # The only diverging change was that we initialize the @cipher and @hmac (for
+    # verification) here instead of in the methods that use them (found in the
+    # `cryptography` base class).
+    #
+    def initialize(key)
+      decoded_key = Base64.urlsafe_decode64(key)
+      if decoded_key.size != 64
+        raise ArgumentError, "Fernet key must be 64 url-safe base64-encoded bytes."
+      end
+
+      @signing_key    = decoded_key[0,32]
+      @encryption_key = decoded_key[32,32]
+
+      @cipher = OpenSSL::Cipher::AES256.new(:CBC)
+      @hmac   = OpenSSL::HMAC.new signing_key, OpenSSL::Digest::SHA256.new
+    end
+
+    # This is from the main bulk of the Fernet#decrypt method in the
+    # cryptography python lib:
+    #
+    #   https://github.com/pyca/cryptography/blob/ed5a7717/src/cryptography/fernet.py#L105-L122
+    #
+    # The verify bit will be described below, but the bluk of what is in the
+    # cryptography lib can simply be replaced with the cipher decrypt
+    # functionality in Ruby's OpenSSL implementation.
+    #
+    # Since the fernet lib is also in charge of a decent amount of FFI
+    # interplay that the OpenSSL equivalent in Ruby takes care of for us in
+    # it's C implementation directly, so if this looks a bit simpler, that is
+    # because it is.
+    #
+    # Of note from the above:  Because of how the string slicing works in
+    # python and differs from the notation in ruby, there looks to be "off by
+    # one" errors all over the place, but that is intentional.  For example,
+    # for the `cipher.iv` assignment, in python it is:
+    #
+    #   iv = data[9:25]
+    #
+    # Which translates to "start at the 9th char, and bring back everything to
+    # the 25 char", where in ruby it is:
+    #
+    #   iv = data[9,16]
```

Comment:
Ruby can do `data[9..25]`, so I'm curious why you chose to use the `,` notation instead?

Reply:
It was 12am this morning and what made sense at the time?

Slightly more accurate/serious answer:  mostly to avoid a `range` object allocation and instead use just integers... #overOptimizingMmmkay

Reply:
For what it is worth, the reason why I did switch between is the `[start:end]` syntax for this in python ("slicing" I think it is called) is a little more flexible than the ruby `[start, length]` syntax.

In python, you are able to omit either the `start` or the `end` values, and it assume the beginning or the end of the string.  In Ruby, that isn't available in this form, and you only can simulate the functionality with the `[start..end]` syntax.  That said, `start` and `end` are still required for a range value, you just are allowed to use negative numbers.

The `ciphertext = data[25..-33]` is a better example of this, since the python equivalent was `ciphertext = data[25:-32]`, so a bit of a minor "off by one" difference in implementation even when using a "more similar" interface.

Reply:
Thanks for the explanation.

> start and end are still required for a range value

FYI, soon...

2.6 - https://ruby-doc.org/core-2.6/Range.html#class-Range-label-Endless+Ranges
2.7 - https://ruby-doc.org/core-2.7.0.preview1/Range.html#class-Range-label-Beginless-2FEndless+Ranges